### ISSUE TITLE
off-the-mainthread fallocate

### ISSUE THREAD

**Issue by nmaier** [aria2/aria2#953]
While aria2 is event-based, the file-allocation code is not. This matters in particular for the fallocate variant, which still might take a considerable amount of time to allocate extents.
The reason this bothers me in the first place: fallocate blocking the main thread for long periods will make my RPC web interface time out a lot.

I got a preliminary patch based on `std::thread` and `std::atomic` to do allocations in a thread, and it seems to work on my linux and macos boxes (didn't try windows yet).

However, adding threads to aria2 is a somewhat big deal, so I wanted to check first if this is something you might wanna do, @tatsuhiro-t, maybe along with a `--disable-threads` build-time option.

**Comment by tatsuhiro-t** [comment]
Modern file system performs fallocate very fast.  It looks like you are talking about the case where user accidentally use --file-allocation=falloc against file system which does not support it (e.g., vfat)?
For those legacy file system, --file-allocation=prealloc should work, and its does not stop the world.

**Comment by nmaier** [comment]
no, I am actually talking about fallocate on ext4 which supports fallocate. The larger and high use, the longer it will take the file system to figure out where to place extents (the max extent size on ext4 is 160MB or something like that, so a multi-megabyte file will take many even without fragmentation, and if the file system is in high use, the number of extents required may increase even more due to fragmentation).

I'm running aria2 on an i686 Atom SoC NAS, and it might sometimes take up to a minute to fallocate a multi-gigabyte file there.

**Comment by tatsuhiro-t** [comment]
Thank you for sharing this.  I haven't used aria2 in such a condition, and you know more about it than me.
I'm fine to add this feature as long as it works reliably (it's a thread!).
